### PR TITLE
tests.buildFHSEnv.libtinfo: init

### PR DESCRIPTION
**Pull request by doronbehar**
## Description of changes

This PR adds two tests to `pkgs/test/` that test whether `buildFHSEnv` actually does what it promises to do - make binaries from the outside world find shared objects in the FHS environment.

I wrote these tests in order to prove that there is something wrong with it - one fails, the other succeeds. I'm not sure exactly what is the problem so I'd like investigate this together with others. Hence I'm ccing @Atemu as it seems you maintain `buildFHSEnv`.

To run the tests:

```
nix-build -A tests.buildFHSEnv.liblzma # not failing
nix-build -A tests.buildFHSEnv.libtinfo # Very similar concept, failing
```

I encountered this issue with the `libtinfo` while trying to use the badly designed, very NixOS-unfriendly tool called [Petalinux](https://www.xilinx.com/products/design-tools/embedded-software/petalinux-sdk.html). The two shared objects committed here were extracted from Petalinux' installer.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/test/buildFHSEnv/liblzma/libxml2.so
Comment (Atemu) [inline pkgs/test/buildFHSEnv/liblzma/libxml2.so:1]: We won't have MiB or even hundreds of KiB of binary data inside the nixpkgs tree. Please fetch these from wherever you got them from using a regular fetcher.

File: pkgs/test/buildFHSEnv/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+}:
+
+let
+  makeSharedObjectTest = sharedObject: targetPkgs: let
+    lddFHSEnv = buildFHSEnv {
+      name = "ldd-with-ncurses-FHS-env";
+      inherit targetPkgs;
+      runScript = "ldd";
+    };
+    ldd-in-FHS = "${lddFHSEnv}/bin/${lddFHSEnv.name}";
```
Comment (Atemu) [inline pkgs/test/buildFHSEnv/default.nix:15]: This is fine but we should probably just provide buildFHSEnv outputs a meta.mainProgram that internalises this logic.

File: pkgs/test/buildFHSEnv/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  '';
+
+in {
+  libtinfo = makeSharedObjectTest ./libtinfo/libedit.so (p: [
+    p.ncurses5
+  ]);
+
+  liblzma = makeSharedObjectTest ./liblzma/libxml2.so (p: [
+    p.xz
+  ]);
```
Comment (Atemu) [inline pkgs/test/buildFHSEnv/default.nix:48]: Please add a comment on the failing one in order to not confuse anyone trying to run it.

**Review by Atemu — CHANGES_REQUESTED** [review]
I don't know where to begin with this issue either but reproducible test cases is much appreciated progress towards solving it.

Adding binaries to the Nixpkgs tree is a big no-no however.

**Comment by Atemu** [comment]
Also note that this is a known problem and not necessarily an FHSEnv one. I don't remember the exact details but I think it's https://github.com/NixOS/nixpkgs/issues/89769.

File: pkgs/test/buildFHSEnv/default.nix (hunk 1 above)
Comment (doronbehar) [reply pkgs/test/buildFHSEnv/default.nix:34]: > This is fine but we should probably just provide buildFHSEnv outputs a meta.mainProgram that internalises this logic.

I'm not sure I understand - do you want to use:

```nix
"${lddFHSEnv}/bin/${lddFHSEnv.meta.mainProgram}"
```

?

**Comment by doronbehar** [comment]
BTW Debian packages `libtinfo` and `libncurses` in two different packages:

https://packages.debian.org/bookworm/amd64/libtinfo6/filelist

https://packages.debian.org/bookworm/amd64/libncurses6/filelist

And we have many symlinks there instead:

```
total 756
lrwxrwxrwx 1 root root     11 Jan  1  1970 libform.so -> libformw.so
lrwxrwxrwx 1 root root     13 Jan  1  1970 libform.so.6 -> libformw.so.6
lrwxrwxrwx 1 root root     13 Jan  1  1970 libformw.so -> libformw.so.6
lrwxrwxrwx 1 root root     15 Jan  1  1970 libformw.so.6 -> libformw.so.6.4
-r-xr-xr-x 1 root root  92064 Jan  1  1970 libformw.so.6.4
lrwxrwxrwx 1 root root     11 Jan  1  1970 libmenu.so -> libmenuw.so
lrwxrwxrwx 1 root root     13 Jan  1  1970 libmenu.so.6 -> libmenuw.so.6
lrwxrwxrwx 1 root root     13 Jan  1  1970 libmenuw.so -> libmenuw.so.6
lrwxrwxrwx 1 root root     15 Jan  1  1970 libmenuw.so.6 -> libmenuw.so.6.4
-r-xr-xr-x 1 root root  50440 Jan  1  1970 libmenuw.so.6.4
lrwxrwxrwx 1 root root     14 Jan  1  1970 libncurses.so -> libncursesw.so
lrwxrwxrwx 1 root root     16 Jan  1  1970 libncurses.so.6 -> libncursesw.so.6
lrwxrwxrwx 1 root root     16 Jan  1  1970 libncursesw.so -> libncursesw.so.6
lrwxrwxrwx 1 root root     18 Jan  1  1970 libncursesw.so.6 -> libncursesw.so.6.4
-r-xr-xr-x 1 root root 517560 Jan  1  1970 libncursesw.so.6.4
lrwxrwxrwx 1 root root     12 Jan  1  1970 libpanel.so -> libpanelw.so
lrwxrwxrwx 1 root root     14 Jan  1  1970 libpanel.so.6 -> libpanelw.so.6
lrwxrwxrwx 1 root root     14 Jan  1  1970 libpanelw.so -> libpanelw.so.6
lrwxrwxrwx 1 root root     16 Jan  1  1970 libpanelw.so.6 -> libpanelw.so.6.4
-r-xr-xr-x 1 root root  25856 Jan  1  1970 libpanelw.so.6.4
lrwxrwxrwx 1 root root     14 Jan  1  1970 libtinfo.so -> libncursesw.so
lrwxrwxrwx 1 root root     16 Jan  1  1970 libtinfo.so.6 -> libncursesw.so.6
lrwxrwxrwx 1 root root     17 Jan  1  1970 terminfo -> ../share/terminfo
```

Could it be that because `libtinfo.so.6` is a symlink `buildFHSEnv` doesn't put it in `/lib/`?

**Comment by Mic92** [comment]
One issue with our ncurses package is that we put the unicode-enabled variant and the non-unicode variant in the same package. Does you test works better if you do this? https://github.com/nix-community/nix-environments/blob/6786fe88f0a45ed42417462b6cd4017172ce9802/envs/xilinx-vitis/shell.nix#L25

File: pkgs/test/buildFHSEnv/default.nix (hunk 1 above)
Comment (Atemu) [reply pkgs/test/buildFHSEnv/default.nix:34]: Ah, indeed `mainProgram` is only the exe name; I got confused.

Though thinking about it again, with  `mainProgram` we could use `getExe` here. That's a discussion for another PR though.

**Comment by doronbehar** [comment]
Thanks @Mic92 for joining the discussion! Sorry for being slow to respond, only now I got into the computer where the proprietary software is located. Indeed the solution you propose is working for the shared object I initially encountered the issue with. I also managed to fix the test!

What do you think of creating packages like these:

```nix
  libtinfo = ncurses.override {
    unicodeSupport = false;
    termlibSupport = true;
  };
  libtinfo5 = ncurses5.override {
    unicodeSupport = false;
    termlibSupport = true;
  };
```

After also making the appropriate modifications to ncurses' expression?

**Comment by doronbehar** [comment]
Disclaimer: I also encountered this issue with that shared object after trying to run a xilinx product - `petalinux` (still without success though, even with this new ncurses knowledge, see [this support thread](https://support.xilinx.com/s/question/0D54U00007tNuYnSAK/petalinux-fails-to-detect-my-gcc)).

**Review by Atemu — APPROVED** [review]
I think it's good to have this even with the failing test and the diff LGTM now.

**Comment by doronbehar** [comment]
Thanks for the approval. Interestingly, the aarch64-linux build fails, and I don't understand the error there...

In anycase, I'd like to get an approval for [the idea proposed in my previous comment](https://github.com/NixOS/nixpkgs/pull/279844#issuecomment-1893262199), as it'd allow to write these tests much more cleanly.

**Comment by Atemu** [comment]
aarch64-linux isn't really supported by buildFHSEnv.

I think it'd be more appropriate to ask for feedback on the idea in the dedicated bug ticket rather than this PR.

**Pull request merged**